hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelinePlacementPolicy.java

```diff
@@ -315,6 +314,50 @@ DatanodeDetails fallBackPickNodes(
     return results;
   }
 
+  private DatanodeDetails randomPick(List<DatanodeDetails> healthyNodes) {
+    DatanodeDetails datanodeDetails;
+    int firstNodeNdx = getRand().nextInt(healthyNodes.size());
+    int secondNodeNdx = getRand().nextInt(healthyNodes.size());
+
+    // There is a possibility that both numbers will be same.
+    // if that is so, we just return the node.
+    if (firstNodeNdx == secondNodeNdx) {
+      datanodeDetails = healthyNodes.get(firstNodeNdx);
+    } else {
+      DatanodeDetails firstNodeDetails = healthyNodes.get(firstNodeNdx);
+      DatanodeDetails secondNodeDetails = healthyNodes.get(secondNodeNdx);
+      datanodeDetails = nodeManager.getPipelinesCount(firstNodeDetails)
+          >= nodeManager.getPipelinesCount(secondNodeDetails)
+          ? secondNodeDetails : firstNodeDetails;
+    }
+    return datanodeDetails;
+  }
+
+  private List<DatanodeDetails> getLowerLoadNodes(
+      List<DatanodeDetails> nodes, int num) {
+    int maxPipelineUsage = nodes.size() * heavyNodeCriteria /
+        HddsProtos.ReplicationFactor.THREE.getNumber();
+    return nodes.stream()
+        // Skip the nodes which exceeds the load limit.
+        .filter(p -> nodeManager.getPipelinesCount(p) < num - maxPipelineUsage)
+        .collect(Collectors.toList());
+  }
+
+  private DatanodeDetails lowerLoadPick(List<DatanodeDetails> healthyNodes) {
+    int curPipelineCounts =  stateManager
+        .getPipelines(HddsProtos.ReplicationType.RATIS).size();
+    DatanodeDetails datanodeDetails;
+    List<DatanodeDetails> nodes = getLowerLoadNodes(
+        healthyNodes, curPipelineCounts);
+    if (nodes.isEmpty()) {
+      // random pick node if nodes load is at same level.
+      datanodeDetails = randomPick(healthyNodes);
+    } else {
+      datanodeDetails = nodes.stream().findFirst().get();
```

Comment:
Could this be simplified to just `nodes.get(0)`? In the else branch we already know the list is not empty, so I think we can just pick the first node in the list safely.

Reply:
Actually, thinking about this more - should this pick a random node from LowerLoadNodes, rather than the first node? We call `getLowerLoadNodes` which returns a list of nodes where the overloaded nodes are removed, but I don't think the list is sorted in anyway. The healthyNodes will will probably be naturally in the same order each time it is generated from the NodeManager. This means this lowerLoadPick method might return the same node on each call until it is overloaded. Then it would be excluded and the next node would be picked and so on. It would probably be better if we picked a random node from the less loaded nodes. Or, sort the list by load and return the first one so we are always picked the node with the least node.

Reply:
When i fixed the suspected bug I mentioned above, and then ran the test. The nodes do appear to fill their piplines on a node by node basis and then the test failed as each node did not have at least the average pipelines.

Making this change got it to pass again:

```
datanodeDetails = nodes.get(getRand().nextInt(nodes.size()) );//stream().findFirst().get();
```

But it might be ever better if we sorted the list by pipeline count ascending and then took the first one, but it would be more expensive.

Reply:
Sorting the node list would be expensive for large cluster. That's the reason why I choose to do this 'water mark' filter for selecting nodes with lower load. 

I can def do a findAny() kinda thing for random pick.

Reply:
You could probably do a "top N" type of sort where you only keep the lowest loaded node, but it would require a bit of a refactor of `getLowerLoadNodes`, probably changing it to `getLowestLodeNode`. That would avoid a full sort and would not be much more expensive that the current code, eg:
```
DatanodeDetails lowest = null;
int lowestPipelineCount;
for (DatanodeDetails dn : nodes) {
  int nodePipelines = node.getPipelineCount();
  if (nodePipelines > limit) {
    continue;
  }
  if ((lowest == null) || lowestPipelineCount > nodePipelines {
    lowest = node;
    lowestPipelineCount = nodePipelines;
  }
}
return lowest;
```

Reply:
I feel like the outcome will be similar tho. The current implementation should work just fine.

Reply:
I am fine with doing a "random pick" from the `getLowerLoadNodes` list or using the idea I had above to get the node with the lowest load each time. I think there are advantages to each of them. I believe we do need to go with one of those ideas, as picking the first one will not work well.

The random pick is simple, but it may not spread the load evenly every time.

Picking the lowest one each time is slightly more complicated, but it does guarantee to always use the lowest load node first and will spread the load evenly for sure. However it is less random - eg if a new node joins the cluster, then it will be used for the next N pipelines until it reaches the same load as some others.

Reply:
@sodonnel I'm navigating thru another problem with chooseNodeFromNetworkTopology. Because topology only gives interface like networkTopology.chooseRandom(anchor.getNetworkLocation(), excluded), I have to move all higher load nodes into excluded and allow topology to pick one. So the original water mark cannot guarantee full load balance due to fall back logic I added. 

I could see your proposed sort will help picking the lowEST node. How could it help leverage to have higher load with topology? I'm still experimenting...